src/celeritas/user/detail/StepDiagnosticImpl.hh

```diff
@@ -0,0 +1,57 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2023 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/user/detail/StepDiagnosticImpl.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "corecel/Assert.hh"
+#include "corecel/Macros.hh"
+#include "corecel/math/Algorithms.hh"
+#include "corecel/math/Atomics.hh"
+#include "celeritas/global/CoreTrackView.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Collect distribution of steps per track for each particle type.
+ */
+inline CELER_FUNCTION void
+tally_steps(CoreTrackView const& track,
+            NativeCRef<ParticleTallyParamsData>&& params,
+            NativeRef<ParticleTallyStateData>&& state)
+{
+    using BinId = ItemId<size_type>;
+
+    CELER_EXPECT(params);
+    CELER_EXPECT(state);
```

Comment:
I think precondition assertions should take preference over the `using` declarations:
```suggestion
    CELER_EXPECT(params);
    CELER_EXPECT(state);

    using BinId = ItemId<size_type>;
```